test/Streamly/Test/Internal/FileSystem/Event.hs

```diff
@@ -0,0 +1,314 @@
+import Test.Hspec
+import Test.Hspec.QuickCheck
+import Control.Monad.IO.Class (MonadIO)
+import Control.Concurrent
+import Data.List.NonEmpty (NonEmpty)
+import Data.Word (Word8)
+import System.Directory
+import System.FilePath ((</>))
+import Streamly.Prelude (SerialT)
+import Streamly.Internal.Data.Array.Storable.Foreign (Array)
+
+import qualified Streamly.Unicode.Stream as Unicode
+import qualified Streamly.Internal.Data.Array.Storable.Foreign as Array
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Streamly.Internal.Data.Stream.IsStream as Stream
+#if defined(CABAL_OS_DARWIN)
+import qualified Streamly.Internal.FileSystem.Event.Darwin as Event
+#elif defined(CABAL_OS_LINUX)
+import qualified Streamly.Internal.FileSystem.Event.Linux as Event
+#elif defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.FileSystem.Event.Windows as Event
+#else
+#error "FS Events not supported on this platform
+#endif
+
+import System.IO
+import qualified Data.Set as Set
+#if !defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.Unicode.Stream as U
+import Data.Functor.Identity (runIdentity)
+#endif
+
+main :: IO ()
+main = hspec $ do
+    prop "Recursive directory events validator" validator
+    prop "Staging directories cleanUp" cleanUp   
```

Comment:
Divide the testing into many small test cases. One test case would be a small sequence of file system operations and the resulting events that are expected from these operations. Now let's write a single test case first:

1. Let's say the function to perform fs ops is called `fsOps`, this would be an IO function. It would perform a sequence of ops under our `watchRootDir`.

```
fsOps synch = do
    takeMVar synch
     ....
```

2. Write a parser `parseEvents` for this test case that would parse the stream of events from `watchRootDir`. The parser would expect a certain sequence of events. Run the parser on the event stream e.g. `checkEvents synch = S.parse parseEvents (S.before (putMVar synch) inputStream)`. Instead of a `Parser`, we could even use `S.take n` to take `n` events from the stream. We already know how much `n` should be based on the test case. And then we can match the list of these events with expected events.

3. Now compose `fsOps` and `checkEvents` to run in parallel using

```
timeout = threadDelay 5000000 >> return (Just Failure)
checker synch = S.yieldM (checkEvents synch) `S.parallel` S.yieldM timeout

 S.head $ checker synch `Par.parallelFst` S.yieldM (fsOps synch)
```

`checkEvents` can return a `Just Success` or `Just Failure` depending on whether the parse succeeded or failed. You can test the value returned by `S.head` to check if the test passed or failed. If we are not able to parse in a given time then we will receive a failure from the timeout thread.

You can replicate this for each test case and write as many tests as you want.